wgpu/gui/qt.py

```diff
@@ -13,50 +14,109 @@
+except Exception:
+    pass  # fail on older Qt's
 
-class QtWgpuCanvas(BaseCanvas, QWidget):
-    """ A QWidget subclass that can be used as a canvas to render to.
-    """
 
+class QtWgpuCanvas(BaseCanvas, QtWidgets.QWidget):
     def __init__(self, *args, size=None, title=None, **kwargs):
         super().__init__(*args, **kwargs)
-        self.setAttribute(QtCore.Qt.WA_PaintOnScreen, True)
-        self.setAutoFillBackground(False)
 
         if size:
             width, height = size
             self.resize(width, height)
+        elif "parent" not in kwargs:
+            self.resize(640, 480)
         if title:
             self.setWindowTitle(title)
-        self.show()
 
-    def paintEngine(self):  # noqa: N802 - this is a Qt method
-        # https://doc.qt.io/qt-5/qt.html#WidgetAttribute-enum  WA_PaintOnScreen
-        return None
+        # The actual surface is held by a subwidget. This is to make sure that
+        # the logical surface size is actually integer. Otherwise the window
+        # size can be set to subpixel (logical) values, without being able to
+        # detect this. See https://github.com/almarklein/wgpu-py/pull/68
+        self._subwidget = WgpuSubWidget(self)
 
-    def paintEvent(self, event):  # noqa: N802 - this is a Qt method
-        self._draw_frame_and_present()
+        layout = QtWidgets.QHBoxLayout(self)
+        layout.setContentsMargins(0, 0, 0, 0)
+        self.setLayout(layout)
+        layout.addWidget(self._subwidget)
 
-    def get_size_and_pixel_ratio(self):
-        return self.width(), self.height(), self.window().devicePixelRatio()
+        self.show()
 
-    def is_closed(self):
-        return not self.isVisible()
+    # Qt methods
+
+    def update(self):
+        super().update()
+        self._subwidget.update()
+
+    # Methods that we add from wgpu (snake_case)
 
     def get_display_id(self):
         # There is qx11info, but it is rarely available.
         # https://doc.qt.io/qt-5/qx11info.html#display
         return super().get_display_id()  # uses X11 lib
 
     def get_window_id(self):
-        return int(self.winId())
+        return int(self._subwidget.winId())
+
+    def get_pixel_ratio(self):
+        # The pixel ratio always seems to be a whole number. When setting
+        # the scale in Windows 10 to 175%, Qt pretends its 2.0.
+        return self._subwidget.devicePixelRatioF()
+
+    def get_logical_size(self):
+        # Sizes in Qt are logical
+        lsize = self._subwidget.width(), self._subwidget.height()
+        return float(lsize[0]), float(lsize[1])
+
+    def get_physical_size(self):
+        # https://doc.qt.io/qt-5/qpaintdevice.html
+        # https://doc.qt.io/qt-5/highdpi.html
+        lsize = self._subwidget.width(), self._subwidget.height()
+        lsize = float(lsize[0]), float(lsize[1])
+        ratio = self._subwidget.devicePixelRatioF()
+        return int(lsize[0] * ratio + 0.4999), int(lsize[1] * ratio + 0.4999)
```

Comment:
is this a rounding trick?

Reply:
Yeah. What was I thinking? Changed to use `round()`.